edx_ace/channel/push_notification.py

```diff
@@ -0,0 +1,107 @@
+        """
+        Transmit a rendered message to a recipient.
+
+        Args:
+            message: The message to transmit.
+            rendered_message: The rendered content of the message that has been personalized
+                for this particular recipient.
+        """
+        device_tokens = self.get_user_device_tokens(message.recipient.lms_user_id)
+        if not device_tokens:
+            LOG.info('Recipient %s has no push token. Skipping push notification.', message.recipient.email_address)
+            return
+
+        for token in device_tokens:
+            self.send_message(message, token, rendered_message)
+
+    def send_message(self, message: Message, token: str, rendered_message: RenderedPushNotification) -> None:
+        """
+        Send a push notification to a device by token.
+        """
+        notification_data = {
+            'title': self.sanitize_html(rendered_message.subject),
+            'body': self.sanitize_html(rendered_message.body),
+            'notification_key': token,
+            **message.context.get('push_notification_extra_context', {}),
+        }
+        message = dict_to_fcm_message(notification_data)
+        # Note: By default dict_to_fcm_message does not support APNS configuration,
+        #   only Android configuration, so we need to collect and set it manually.
+        apns_config = self.collect_apns_config(notification_data)
+        message.apns = apns_config
+        try:
+            send_message(token, message, settings.FCM_APP_NAME)
+        except Exception as e:
+            LOG.exception('Failed to send push notification to %s', token)
+            raise FatalChannelDeliveryError('Failed to send push notification to {token}') from e
+
+    @staticmethod
+    def collect_apns_config(notification_data: dict) -> APNSConfig:
+        """
+        Collect APNS configuration with payload for the push notification.
+
+        This APNSConfig must be set to notifications for Firebase to send push notifications to iOS devices.
+        Notification has default priority and visibility settings, described in Apple's documentation.
+        (https://developer.apple.com/documentation/usernotifications/sending-notification-requests-to-apns)
+        """
+        apns_alert = ApsAlert(title=notification_data['title'], body=notification_data['body'])
+        aps = Aps(alert=apns_alert, sound='default')
+        return APNSConfig(
+            headers={'apns-priority': APNS_DEFAULT_PRIORITY, 'apns-push-type': APNS_DEFAULT_PUSH_TYPE},
+            payload=APNSPayload(aps)
+        )
+
+    @staticmethod
+    def get_user_device_tokens(user_id: int) -> list:
+        """
+        Get the device tokens for a user.
+        """
+        return list(GCMDevice.objects.filter(
+            user_id=user_id,
+            cloud_message_type='FCM',
+            active=True,
+        ).values_list('registration_id', flat=True))
+
+    @staticmethod
+    def sanitize_html(html_str: str) -> str:
```

Comment:
Why this method is needed?

Also, `sanitize_html` indicates that this method prevents XSS, but its job is mostly to compress spaces. 

```suggestion
    def cleanup_spaces(html_str: str) -> str:
```

or

```suggestion
    def compress_spaces(html_str: str) -> str:
```

Reply:
Renamed